control/endpoints.go

```diff
@@ -11,70 +11,66 @@ import (
 	"github.com/joyent/containerpilot/events"
 )
 
-// Endpoints defines bridge data that cross the App and HTTPServer API boundary
+// Endpoints wraps the EventBus so we can bridge data across the App and
+// HTTPServer API boundary
 type Endpoints struct {
 	bus *events.EventBus
 }
 
 // PostHandler is an adapter which allows a normal function to serve itself and
-// handle incoming HTTP POST requests. Also allows us to pass through App state
-// in a more organized fashion.
-type PostHandler func(http.ResponseWriter, *http.Request)
-
-// ServeHTTP implements intermediate endpoint behavior for POST
-// requests. Subsequently calls actual handler implementations defined below.
-func (ph PostHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	log.Debugf("control: '%s %s' requested", r.Method, r.URL)
+// handle incoming HTTP POST requests, and allows us to pass thru EventBus to
+// handlers
+type PostHandler func(*http.Request) (interface{}, int)
 
+func (pw PostHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	if r.Method != http.MethodPost {
 		failedStatus := http.StatusNotImplemented
-		log.Errorf("%s requires POST, not %s", r.URL, r.Method)
 		http.Error(w, http.StatusText(failedStatus), failedStatus)
 		return
 	}
-
-	w.Header().Set("Content-Type", "application/json")
-	w.WriteHeader(http.StatusOK)
-
-	ph(w, r)
-
-	io.WriteString(w, "\n")
+	resp, status := pw(r)
+	switch status {
+	case http.StatusOK:
+		if resp != nil {
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+			json.NewEncoder(w).Encode(resp)
+		} else {
+			w.WriteHeader(http.StatusOK)
+			io.WriteString(w, "\n")
```

Comment:
It's something to watch out for, for sure. In previous web APIs I've done this as composable functions so that I end up with formatter functions (ex. content type renderers) that wrap the switch-on-error that wrap the underlying endpoint. Most of the golang 3rd party routers take this approach.